### PR TITLE
[FLINK-28327] Make table store src codes compiled with Flink 1.14

### PR DESCRIPTION
**Pull request by JingsongLi**
- Copy incompatible codes from Flink to Table Store
- Provide different versions for 1.15 and 1.14
- Add action: Build Flink 1.14

### CODE REVIEW

File: flink-table-store-codegen-loader/src/main/java/org/apache/flink/table/store/codegen/CodeGenLoader.java [hunk 1]
```diff
@@ -58,18 +63,6 @@ public class CodeGenLoader {
 
     private static final String[] COMPONENT_CLASSPATH = new String[] {"org.apache.flink"};
 
-    private static final Map<String, String> KNOWN_MODULE_ASSOCIATIONS = new HashMap<>();
```
Comment (JingsongLi) [inline flink-table-store-codegen-loader/src/main/java/org/apache/flink/table/store/codegen/CodeGenLoader.java:61]: Useless, just for log... So remove this.

File: pom.xml [hunk 2]
```diff
@@ -110,6 +110,16 @@ under the License.
         <!-- Can be set to any value to reproduce a specific build. -->
         <test.randomization.seed/>
         <test.unit.pattern>**/*Test.*</test.unit.pattern>
+
```
Comment (LadyForest) [inline pom.xml:113]: I guess there is something wrong during rebase/merge. L#85~L#101 should hold the very first place.

File: flink-table-store-connector/src/main/1.14.5/org/apache/flink/table/store/connector/TableStoreDataStreamScanProvider.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.flink.table.connector.source.DataStreamScanProvider;
+import org.apache.flink.table.data.RowData;
+
+import java.util.function.Function;
+
+/** Table Store {@link DataStreamScanProvider}. */
+public class TableStoreDataStreamScanProvider implements DataStreamScanProvider {
+
+    private final boolean isBounded;
+    private final Function<StreamExecutionEnvironment, DataStream<RowData>> produce;
```
Comment (LadyForest) [inline flink-table-store-connector/src/main/1.14.5/org/apache/flink/table/store/connector/TableStoreDataStreamScanProvider.java:32]: Nit: how about renaming to `producer`?

**Review by LadyForest — CHANGES_REQUESTED** [review]
Thanks for the great work! It looks good in general, and I only left some comments.

**Review by LadyForest — CHANGES_REQUESTED** [review]
Thanks for the great work! It looks good in general, and I only left some comments.

File: .github/workflows/build-flink-1.14.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Build
+        run: |
+          mvn clean install -Dmaven.test.skip=true -Pflink-1.14
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply .github/workflows/build-flink-1.14.yml:18]: Make tests compiled with Flink 1.14 is very very hard...

File: flink-table-store-common/src/main/java/org/apache/flink/table/store/utils/ReflectionUtils.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            if (methodName.equals(m.getName())) {
+                method = m;
+                break;
+            }
+        }
+
+        if (method == null) {
+            throw new NoSuchMethodException(methodName);
+        }
+        return (T) method.invoke(null, args);
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table-store-common/src/main/java/org/apache/flink/table/store/utils/ReflectionUtils.java:42]: We don't want to visit private method.

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java [hunk 6]
```diff
@@ -117,7 +118,7 @@ public Predicate visit(CallExpression call) {
                     allowQuick = true;
                 } else if (escape != null) {
                     if (escape.length() != 1) {
-                        throw SqlLikeUtils.invalidEscapeCharacter(escape);
+                        throw new RuntimeException("Invalid escape character '" + escape + "'");
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/predicate/PredicateConverter.java:121]: I thought about it, but it was too simple and didn't feel necessary

File: flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/RowDataPartitionComputer.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+                        .toArray(RowData.FieldGetter[]::new);
+    }
+
+    public LinkedHashMap<String, String> generatePartValues(RowData in) {
+        LinkedHashMap<String, String> partSpec = new LinkedHashMap<>();
+
+        for (int i = 0; i < partitionFieldGetters.length; i++) {
+            Object field = partitionFieldGetters[i].getFieldOrNull(in);
+            String partitionValue = field != null ? field.toString() : null;
+            if (partitionValue == null || "".equals(partitionValue)) {
```
[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/RowDataPartitionComputer.java:53]: `"".equals(partitionValue)` returns false, but we want true.

[the remark this answers is not in the document]
Comment (JingsongLi) [reply flink-table-store-core/src/main/java/org/apache/flink/table/store/file/utils/RowDataPartitionComputer.java:53]: Replace it with `StringUtils.isNullOrWhitespaceOnly`

**Review by LadyForest — APPROVED** [review]
Thanks for the update, looks good to me!

**Pull request merged**